language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -157,54 +225,42 @@ runCekM
     -> ExBudgetState
     -> CekM uni a
     -> (Either (CekEvaluationException uni) a, ExBudgetState)
-runCekM env s a = runState (runQuoteT . runExceptT $ runReaderT a env) s
+runCekM env s a = runState (runExceptT $ runReaderT a env) s
 
--- | Get the current 'VarEnv'.
-getVarEnv :: CekM uni (VarEnv uni)
-getVarEnv = asks _cekEnvVarEnv
+-- | Get the current 'CekValEnv'.
+getEnv :: CekM uni (CekValEnv uni)
+getEnv = asks _cekEnvVarEnv
 
 -- | Set a new 'VarEnv' and proceed.
-withVarEnv :: VarEnv uni -> CekM uni a -> CekM uni a
-withVarEnv venv = local (set cekEnvVarEnv venv)
+withEnv :: CekValEnv uni -> CekM uni a -> CekM uni a
+withEnv venv = local (set cekEnvVarEnv venv)
 
 -- | Extend an environment with a variable name, the value the variable stands for
 -- and the environment the value is defined in.
-extendVarEnv :: Name -> WithMemory Value uni -> VarEnv uni -> VarEnv uni -> VarEnv uni
-extendVarEnv argName arg argVarEnv =
-    insertByName argName $ Closure argVarEnv arg
+extendEnv :: Name -> CekVal uni -> CekValEnv uni -> CekValEnv uni
+extendEnv argName arg  =
+    insertByName argName arg
 
 -- | Look up a variable name in the environment.
-lookupVarName :: Name -> CekM uni (Closure uni)
+lookupVarName :: Name -> CekM uni (CekVal uni)
 lookupVarName varName = do
-    varEnv <- getVarEnv
+    varEnv <- getEnv
     case lookupName varName varEnv of
-        Nothing   -> throwingWithCause _MachineError
-            OpenTermEvaluatedMachineError
-            (Just . Var (memoryUsage ()) $ varName)
-        Just clos -> pure clos
+        Nothing  -> throwingWithCause _MachineError OpenTermEvaluatedMachineError $ Nothing
```

Comment:
I don't recall what happened to the error term here. Do you?

I was probably rushing to get the stats. We should try restoring the culprit info.

Reply:
The problem is that we have (for example) `MachineError err term` and now `term` is `CekValue`, so we can't put a `Term` into the error.  There's no suitable CekValue for Error, so we're stuck.  The same thing happens in a few other places as well. I'm not sure how to fix that.

Reply:
So the issue is that sometimes we want to give a value in a `MachineError` and sometimes a term? What about always giving a term and just running `discharge` on the value if we need to provide a value? We'd be in an error case so it's okay to be slow.

Reply:
> What about always giving a term and just running discharge on the value if we need to provide a value?

I'm not sure if that's possible.  This is eventually handled by `applyBuiltinName` in `Apply.hs`, which has the type

```
applyBuiltinName
    :: forall m err uni term
    .  ( MonadError (ErrorWithCause err term) m, AsUnliftingError err, AsConstAppError err term
       , SpendBudget m term, HasConstantIn uni term, GShow uni, GEq uni, DefaultUni <: uni
       )
    => BuiltinName -> [term] -> m (ConstAppResult term)
```
It's all parameterised over the type `term`, which in the case of the CEK machine is `CekValue`.  We'd need to return a `Term` in case of an error, but a `CekValue` in case of success.  I'm not sure if you can do that, but maybe there's some monad trickery I haven't spotted.